File: app/src/main/java/com/konektedi/vs/Home/Elections/ElectionsAdapter.java
package com.konektedi.vs.Home.Elections;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.konektedi.vs.R;

import java.util.List;

public class ElectionsAdapter extends RecyclerView.Adapter<ElectionsAdapter.ViewHolder> {

    private Context mContext;
    private List<ElectionsModel> electionList;

    ElectionsAdapter(Context mContext, List<ElectionsModel> electionList) {
        this.mContext = mContext;
        this.electionList = electionList;
    }

    @NonNull
    @Override
    public ElectionsAdapter.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view;
        view = LayoutInflater.from(mContext).inflate(R.layout.z_election_item, parent, false);

        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ElectionsAdapter.ViewHolder holder, final int position) {
        holder.title.setText(electionList.get(position).getElection_title().toUpperCase());


        holder.title.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(mContext, ElectionView.class);
                intent.putExtra("election_id", electionList.get(position).getElection_id());
                intent.putExtra("election_title", electionList.get(position).getElection_title());
                mContext.startActivity(intent);
            }
        });
    }

    @Override
    public int getItemCount() {
        if (electionList != null)
            return electionList.size();
        else return 0;
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        private TextView title;

        ViewHolder(View itemView) {
            super(itemView);

            title = itemView.findViewById(R.id.title);

        }
    }
}

File: app/src/main/java/com/nyererefy/elections/ElectionsAdapter.kt
package com.nyererefy.elections

import androidx.paging.PagedListAdapter
import android.content.Context
import androidx.recyclerview.widget.DiffUtil
import android.view.ViewGroup
import com.nyererefy.R
import com.nyererefy.utilities.common.NetworkState
import com.nyererefy.utilities.common.NetworkStateItemViewHolder
import com.nyererefy.utilities.models.Election

class ElectionsAdapter(private val mContext: Context, private val retryCallback: () -> Unit)
    : PagedListAdapter<Election, androidx.recyclerview.widget.RecyclerView.ViewHolder>(POST_COMPARATOR) {

    private var networkState: NetworkState? = null

    override fun onBindViewHolder(holder: androidx.recyclerview.widget.RecyclerView.ViewHolder, position: Int) {
        when (getItemViewType(position)) {
            R.layout.z_election -> (holder as ElectionsViewHolder).bind(getItem(position), mContext)
            R.layout.network_state_item -> (holder as NetworkStateItemViewHolder).bindTo(
                    networkState)
        }
    }

    override fun onBindViewHolder(
            holder: androidx.recyclerview.widget.RecyclerView.ViewHolder,
            position: Int,
            payloads: MutableList<Any>) {
        if (payloads.isNotEmpty()) {
            val item = getItem(position)
            (holder as ElectionsViewHolder).bind(item, mContext)
        } else {
            onBindViewHolder(holder, position)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): androidx.recyclerview.widget.RecyclerView.ViewHolder {
        return when (viewType) {
            R.layout.z_election -> ElectionsViewHolder.create(parent)
            R.layout.network_state_item -> NetworkStateItemViewHolder.create(parent, retryCallback)
            else -> throw IllegalArgumentException("unknown view type $viewType")
        }
    }

    private fun hasExtraRow() = networkState != null && networkState != NetworkState.LOADED

    override fun getItemViewType(position: Int): Int {
        return if (hasExtraRow() && position == itemCount - 1) {
            R.layout.network_state_item
        } else {
            R.layout.z_election
        }
    }

    override fun getItemCount(): Int {
        return super.getItemCount() + if (hasExtraRow()) 1 else 0
    }

    fun setNetworkState(newNetworkState: NetworkState?) {
        val previousState = this.networkState
        val hadExtraRow = hasExtraRow()
        this.networkState = newNetworkState
        val hasExtraRow = hasExtraRow()
        if (hadExtraRow != hasExtraRow) {
            if (hadExtraRow) {
                notifyItemRemoved(super.getItemCount())
            } else {
                notifyItemInserted(super.getItemCount())
            }
        } else if (hasExtraRow && previousState != newNetworkState) {
            notifyItemChanged(itemCount - 1)
        }
    }

    companion object {
        val POST_COMPARATOR = object : DiffUtil.ItemCallback<Election>() {
            override fun areContentsTheSame(oldItem: Election, newItem: Election): Boolean =
                    oldItem == newItem

            override fun areItemsTheSame(oldItem: Election, newItem: Election): Boolean =
                    oldItem.electionId == newItem.electionId
        }

    }
}
